### DISCUSSION TITLE
Fire JointJS Tooltip from another button

### DISCUSSION THREAD

**Discussion by pandika-csgi** [clientIO/joint#1716] [Q&A] [answered]
Hello, I would ask about, how to make tooltip show and hide from custom event like _state_ in our app, if the _state_ is **off**, hide tooltip, otherwise, if _state_ is **on** show the tooltip

I'm also try used jquery trigger('mouseover') and dispatchevent('mouseover') still not working, is there any other way to do it?

Thanks

**Comment by kumilingus** [comment, marked as answer]
This worked for me.
```js
$el.trigger('mouseenter');
```

**Reply by pandika-csgi** [reply]
> ```js
> trigger('mouseenter')
> ```

Hmm still not work here.
or my selector doesn't select properly, how to get that $el?
I have button with an id and data-tooltip
`$('#btn-layout').trigger('mouseenter')` -> still doesn't work
but when I change tooltip trigger into `click` it worked
`$('#btn-layout').trigger('click')` -> worked

and one more question, if I trigger click inside paper why tooltip doesn't work?

**Reply by pandika-csgi** [reply]
Another Example

```html
<input ... data-attribute="target" data-tooltip="Toggle display" data-tooltip-position="right">
 ```
 `$('[data-attribute="target"]').trigger('mouseenter')` -> still doesn't work

**Reply by kumilingus** [reply]
I opened the KitchenSink app and the following code worked in the browser console (Chrome & FF):
```js
// show the first element tooltip in the stencil
$('.joint-stencil .joint-element').first().trigger('mouseenter')
// show the layout button tooltip
$('#btn-layout').trigger('mouseenter')
```

> and one more question, if I trigger click inside paper why tooltip doesn't work?

What do you think it should do? I don't know the structure of your application. Please share an example of refer to an existing application.

**Reply by kumilingus** [reply]
There is an API to toggle the tooltip visibility. It is not public / documented but you can use it. I think it makes sense to make the methods public in the next JointJS+ release.

```js
const tooltip = new ui.Tooltip({ /* ... */ });
const target = document.querySelector('.my-element-with-tooltip');
tooltip.show({ target });
tooltip.hide({ target })
```

**Reply by pandika-csgi** [reply]
> I opened the KitchenSink app and the following code worked in the browser console (Chrome & FF):
> 
> ```js
> // show the first element tooltip in the stencil
> $('.joint-stencil .joint-element').first().trigger('mouseenter')
> // show the layout button tooltip
> $('#btn-layout').trigger('mouseenter')
> ```
> 
> > and one more question, if I trigger click inside paper why tooltip doesn't work?
> 
> What do you think it should do? I don't know the structure of your application. Please share an example of refer to an existing application.

Ah yeah, I tried in KitchenSink app is worked, I try in console use mouseenter is worked

**Reply by pandika-csgi** [reply]
> There is an API to toggle the tooltip visibility. It is not public / documented but you can use it. I think it makes sense to make the methods public in the next JointJS+ release.
> 
> ```js
> const tooltip = new ui.Tooltip({ /* ... */ });
> const target = document.querySelector('.my-element-with-tooltip');
> tooltip.show({ target });
> tooltip.hide({ target })
> ```

Its work for the `show` method and when { target } I got an error said `Argument of type '{ target: Element; }' is not assignable to parameter of type 'Event'.
  Type '{ target: Element; }' is missing the following properties from type 'TriggeredEvent<any, any, any, any>': currentTarget, delegateTarget, data, bubbles, and 37 more.` I use Angular and rappid v3.4.x, and for `hide` method not working here. 
In my case is need to show all tooltip. 

Sorry, I'm new in this library,
Thanks

**Reply by pandika-csgi** [reply]
Updated if Im only use `show()` and `hide()` is working, is that tooltip API is only show first element? or can do all element target when initialize in `new ui.Tooltip({ target: [data-tooltip-target] /* ... */ });`

Thanks

**Reply by kumilingus** [reply]
You will have to `@ts-ignore` it for now (the `show`/`hide` method expects an event) or create an event programatically.

```js
const target = document.querySelector('.my-element-with-tooltip');
const evt = $.Event('myevent', { target });
tooltip.show(evt);
tooltip.hide(evt);
```

**Reply by pandika-csgi** [reply]
okay but, the hide method `Expected 0 arguments, but got 1`
I'll make that API method as answer even little bit odd in my code and I'll ask about retrieving cell id after on change on graph in new discuss
Thanks @kumilingus

**Reply by kumilingus** [reply]
You are right about `hide()` method.

```js
const target = document.querySelector('.my-element-with-tooltip');
const evt = $.Event('myevent', { target });
tooltip.show(evt);
tooltip.hide();
```

Anyway, we'll fix the typescript definition and add these 2 methods to the documentation ready for the next release.
